### ISSUE TITLE
Using label and arcskip on a link breaks horizontal alignment

### ISSUE THREAD

**Issue by 2pl** [sverweij/mscgen_js#250]
Hi,

I'm having issue with horizontal alignment when using links that have a label attribute and a non zero arcskpip value. The below Xu code is showing two examples of this issue.

I would expect the "Update" link to be always horizontally aligned with the end of the first link, but only the first alternative is showing correct alignment, adding the "Send" label to the first link breaks horizontal alignment.

Tried this with the [msc online](http://mscgen.js.org) and various browser with same result 

```
msc {

kfk2 [label="Kafka"], 
k2db [label="KFK 2 DB"],
mysql [label="MySQL"];

kfk2 alt mysql [label="OK"]{
    kfk2 => k2db [arcskip=1];
    k2db => mysql [label="Update"];
};

kfk2 alt mysql [label="NOT OK"]{
    kfk2 => k2db [label="Send", arcskip=1];
    k2db => mysql [label="Update"];
};


kfk2 alt mysql [label="NOT OK"]{
    kfk2 => k2db [label="Send", arcskip=2];
    k2db => mysql [label="Update"];
};

}
```

**Comment by sverweij** [comment]
Thanks! There's certainly room for improvement there. I'll have to wrap my head around the problem a bit before there'll be a solution 

For rows with an `arcskip` mscgen.js currently takes the base (~ the minimum) height of a row ([source](https://github.com/mscgenjs/mscgenjs-core/blob/develop/render/graphics/renderast.js#L817)). Instead ...
- It should take the real height of the current row
- To be really nice it should also take the height of the next row into account.
(- And to be really _really_ nice it should look for the height of the previous row when `arcskip` < 0).

I also see that even in the 'OK' part of your chart the arc is pointing a pixel too high. I'll fix that too.

![screen shot 2017-11-19 at 10 10 14](https://user-images.githubusercontent.com/4822597/32989242-dd026786-cd12-11e7-82d1-41a5732c3943.png)

**Comment by sverweij** [comment]
I'm working on a fix (☝️). It already solves part of the problem, and addresses some other inconsistencies with the arcsckips and arc gradients (see the PR for details).

I'll have to see if I can tackle (part of) the remaining stuff now or release this first and tackle the rest it later.

**Comment by sverweij** [comment]
@2pl I've published the improvements the previous comment talks about (to npm, to mscgen.js.org and to atom.io for atom-mscgen-preview). 

Deep link with the chart in the interpreter [here](https://mscgen.js.org/?lang=xu&style=lazy&msc=msc%20%7B%0A%0Akfk2%20%5Blabel%3D%22Kafka%22%5D%2C%20%0Ak2db%20%5Blabel%3D%22KFK%202%20DB%22%5D%2C%0Amysql%20%5Blabel%3D%22MySQL%22%5D%3B%0A%0Akfk2%20alt%20mysql%20%5Blabel%3D%22OK%22%5D%7B%0A%20%20%20%20kfk2%20%3D%3E%20k2db%20%5Barcskip%3D1%5D%3B%0A%20%20%20%20k2db%20%3D%3E%20mysql%20%5Blabel%3D%22Update%22%5D%3B%0A%7D%3B%0A%0Akfk2%20alt%20mysql%20%5Blabel%3D%22OK%20(check%3F)%22%5D%7B%0A%20%20%20%20kfk2%20%3D%3E%20k2db%20%5Blabel%3D%22Send%22%2C%20arcskip%3D1%5D%3B%0A%20%20%20%20k2db%20%3D%3E%20mysql%20%5Blabel%3D%22Update%22%5D%3B%0A%7D%3B%0A%0A%0Akfk2%20alt%20mysql%20%5Blabel%3D%22NOT%20OK%20(but%20a%20little%20better%20then%20before%2C%20check%3F)%22%5D%7B%0A%20%20%20%20kfk2%20%3D%3E%20k2db%20%5Blabel%3D%22Send%22%2C%20arcskip%3D2%5D%3B%0A%20%20%20%20k2db%20%3D%3E%20mysql%20%5Blabel%3D%22Update%22%5D%3B%0A%7D%3B%0A%0A%7D); your chart now looks like this:
<img width="501" alt="screen shot 2017-11-19 at 17 56 06" src="https://user-images.githubusercontent.com/4822597/32993059-25aed8d0-cd53-11e7-988e-095223d32453.png">

Still not perfect, but better then it was. 

(B.t.w. For comparison I've taken a look at how the mscgen reference implementation would've rendered this (minus the xù specific stuff):
- ![original mscgen output](https://user-images.githubusercontent.com/4822597/32993333-da484602-cd56-11e7-8cfa-ea7ea437c384.png) ... which looks just as odd as mscgenjs' implementation, but when we add an arc it makes sense again:
- ![x](https://user-images.githubusercontent.com/4822597/32993383-9289e91e-cd57-11e7-91b4-1d298c1174d3.png))

**Comment by 2pl** [comment]
@sverweij thank you for the fix, it definitely improves rendering for arcskip=1 with label use case !

As for the issue with arcskip=2, it is indeed not related to label, I found other examples of weird alignement when using arcskip=2, it looks like using arcksip=2 is shifting the source line down, see below

```mscgen
msc {

kfk2 [label="Kafka"], 
k2db [label="KFK 2 DB"],
mysql [label="MySQL"];

kfk2 alt mysql [label="This is now OK !"]{
    kfk2 >> mysql, 
    kfk2 => k2db [label="Send", arcskip=1];
    k2db => mysql [label="Update"];
};

kfk2 alt mysql [label="not OK"]{
    kfk2 >> mysql, 
    kfk2 => k2db [arcskip=1];
    k2db => mysql [label="Update"];
    k2db => mysql [label="Update"],
    kfk2 >> k2db [label="skip 1 is now ok", arcskip=1];
    k2db => mysql [label="Update"];
    k2db => mysql [label="I feel a bit low"],
    kfk2 >> k2db [label="skip 2 is not ok", arcskip=2];
    k2db => mysql [label="Update"];
};

kfk2 alt mysql [label="out of the box !"]{
    k2db => mysql [label="Update"];
    k2db => mysql [label="I should be higher"],
    kfk2 >> k2db [label="oops skipping out of the box", arcskip=2];
};

kfk2 alt mysql [label="not OK even with no label"]{
    kfk2 >> mysql [label="I am a bit low"], 
    kfk2 >> k2db [arcskip=2], 
    kfk2 => k2db [arcskip=1];
    k2db => mysql [label="I look a bit high but i'm OK !"];
};
}
```

**Comment by sverweij** [comment]
One step at a time :-) 

To be transparent: I work on a thing and if it's an improvement release it (☝️  to witness) instead of making a complete solution covering all cases. This allows me to work on it during the small slots of time I have available and still progress. Otherwise weeks would go by before there's movement on an issue.

E.g. the PR above makes that `arcskip` for _n > 1_ behaves more as expected. I released it yesterday (mscgenjs 1.12.6) It does not solve cases where the height of an arc is larger than the default and it does not take into account inline expressions - which I'll work on next (in that order).

**Comment by sverweij** [comment]
@2pl TL;DR - the remaining issues are now merged and published (to all npm libraries, mscgen.js.org and the atom plugin); give it a spin and let me know what you think.

Thanks again for raising this; it has made the arcskip functionality a lot better than it was; more predictable and (I think) better looking!

Details are in the PR's (so if you feel like it peruse them at will), but I made two decisions that might need some 'splaining:
1. **pointing out of the box**    
For inline expressions (technical name for those `alt`, `opt` & `loop` things) I had to take a decision:
    - should we point to the edge of the inline expression (which is meaningless) ...
    - or should we point it to the next 'real' arc row
The _road of least astonishment_ is the second one; it's weird to count edges of inline expressions as arcs in the first place.

<img width="497" alt="arcskip-out-of-the-box" src="https://user-images.githubusercontent.com/4822597/33495255-fb9638d2-d6c5-11e7-8638-ffd332f8b95c.png">


2. **to infinity... and beyond**    
What to do with arcskips that skip beyond the chart?    
I've followed the original MscGen implementation: max out at the last arc row. In this example chart, `arcskip=42` doesn't get us further than the next arc, because that's the last in the chart:
<img width="284" alt="and-beyond" src="https://user-images.githubusercontent.com/4822597/33495395-8ce90904-d6c6-11e7-9e87-270c55817316.png">

(_B.t.w.; the correct way to point lower than that would be to add an empty row (`|||`, or even `...`) on the bottom_).
